WB-1052 - Add `validation` and `onError` Props to TextField

RivasCVA opened the pull request:
## Summary:
Adding the `validation` and `onError` Props to TextField.
The `validation` prop will be called after every value change.
The `onError` prop will be called after every failed `validation`.

Added unit tests to test both new props.

Included examples on Styleguidist and React Storybook of `validation`
and `onError` working with the `password`, `email`, and `telephone`
type TextFields.

Issue: https://khanacademy.atlassian.net/browse/WB-1052

## Test plan:
1. Open Styleguidist (or React Storybook)
2. Go to the TextField (under Form -> TextField)
3. See the given examples, specially the `password`, `email`, and
   `telephone` TextField examples where `validation` and `onError` are at work.

---

#### Below are examples of validations with error messages:

##### Failed validation on `password` TextField
![tf-password](https://user-images.githubusercontent.com/60367213/120011633-23bc0b00-bfa4-11eb-9579-e637dfd0e69c.png)

##### Failed validation on `email` TextField
![tf-email](https://user-images.githubusercontent.com/60367213/120011778-477f5100-bfa4-11eb-903b-4e5687480bce.png)

#### Failed validation on `telephone` TextField
![tf-telephone](https://user-images.githubusercontent.com/60367213/120011811-5403a980-bfa4-11eb-9d44-c1bc8b43425f.png)

jandrade commented on packages/wonder-blocks-form/src/components/text-field.md:
```diff
@@ -54,31 +54,215 @@ class TextFieldExample extends React.Component {
             value: "Password123",
+            errorMessage: null,
         };
+        this.handleOnChange = this.handleOnChange.bind(this);
+        this.handleValidation = this.handleValidation.bind(this);
+        this.handleOnError = this.handleOnError.bind(this);
```
Ugh this is unfortunate, but hopefully we'll avoid using this once we fully migrate to Storybook 🤞

RivasCVA replied:
🤞

jandrade commented on packages/wonder-blocks-form/src/components/text-field.md:
```diff
@@ -54,31 +54,215 @@ class TextFieldExample extends React.Component {
+        if (!emailRegex.test(value)) {
+            return "Please enter a valid email";
+        }
+        this.setState({errorMessage: null})
```
mmm I'm thinking that this function should only take care of validating the value without having to modify the state (or anything related about React lifecycle and state). I understand that you had to set state here to reset the error state in case the field is valid, but I wonder if we need to modify the `onError` prop to be called no matter if it's an error or not.

That said, I can see different approaches:
1. Modify `onError` to be called every time the validation is performed.
2. Rename `onError` to something more meaningful: e.g. `onValidation(errorMessage: string)`.

IMO using `onValidation` feels more natural in the sense that it's the function that is called after performing `validation`, but ofc I'd like to hear your thoughts on this, and probably explore other possibilities.

jandrade replied:
BTW Thanks for adding this example! It helped to identify this situation ahead of time :)

RivasCVA replied:
I see.... I agree with the `onValidation` route.

At first, I viewed it as redundant since `validation` will call `onValidation`, so why not just put everything in the `validation` function? 

But then thinking deeper about it, I can see why it would be better to have both:
* `validation` will **only** handle validating the value
* `onValidation` will handle or modify any states or variables.

jandrade commented on packages/wonder-blocks-form/src/components/text-field.js:
```diff
@@ -70,8 +81,16 @@ export default class TextField extends React.Component<Props, State> {
+        if (validation) {
+            const isError = validation(newValue);
+            this.setState({error: isError});
+            if (isError && onError) {
+                onError(isError);
+            }
```
Please see my comment related to renaming/calling `onError`.

jandrade reviewed (commented):
Things are looking good, but want to open a conversation about `onError` before continuing with the review.

Thanks for all the examples and tests that you have provided 💯.

RivasCVA commented on packages/wonder-blocks-form/src/components/__tests__/text-field.test.js:
```diff
@@ -130,4 +130,108 @@ describe("TextField", () => {
+        wrapper.simulate("change", {target: {value: newValue}});
+
+        // Assert
+        expect(handleValidation.mock.results[0].value).toEqual(undefined);
```
Actually... I remember now why I used that:
The `toHaveBeenCalledWith` matcher only tests the parameter passed to the `handleValidation` function, but I wanted to test against the **return value** of that function (to ensure it returned `undefined`, meaning a valid input).

jandrade replied:
Got it! thanks for the context. In that case, maybe you could try with `toHaveReturnedWith` and see if it gets the same result?

RivasCVA replied:
It works with `toHaveReturnedWith` 🎉 . Thanks for teaching me of that new matcher 🙏  (I Googled it and no such matcher popped up in my search).

RivasCVA commented on packages/wonder-blocks-form/src/components/__tests__/text-field.test.js:
```diff
@@ -130,4 +130,108 @@ describe("TextField", () => {
+        wrapper.simulate("change", {target: {value: newValue}});
+
+        // Assert
+        expect(handleValidation.mock.results[0].value).toEqual(errorMessage);
```
As commented [above](https://github.com/Khan/wonder-blocks/pull/1128#discussion_r641780627), I want to test the **return value** of the `handleValidation` function, and not the parameter.

Here I want to ensure it returns the string error message that the function is expected to return when the validation fails.

RivasCVA commented:
What changed:
* Replaced all instances of `onError` with `onValidation`. The new prop, `onValidation`, is called every time the input is validated, whether there is an error or not.
* Updated all examples in Styleguidist and React Storybook to use `onValidation` instead of `onError`.
* Updated the unit test handling `onError` to instead test `onValidation`.

jandrade commented on packages/wonder-blocks-form/src/components/text-field.js:
```diff
@@ -70,8 +81,16 @@ export default class TextField extends React.Component<Props, State> {
+        const {validation, onValidation, onChange} = this.props;
+        const newValue = event.target.value;
+        if (validation) {
+            const maybeError = validation(newValue);
```
Great! it feels more consistent

jandrade reviewed (approved):
This looks great! thank you so much for changing to `onValidation`.

One more thing, could you please update the implementation spec to match this new approach?

Great progress!! 🚀

RivasCVA commented:
> This looks great! thank you so much for changing to `onValidation`.
> 
> One more thing, could you please update the implementation spec to match this new approach?
> 
> Great progress!! 🚀

I'm on it 📝

jandrade commented:
I noticed that we missed one case... We need to validate too if the initial value is valid (and if the `validation` prop is set).
Not sure if you have plans to add that in the future, but if not, I think it would be good to add it with this PR.

RivasCVA commented:
> I noticed that we missed one case... We need to validate too if the initial value is valid (and if the `validation` prop is set).
> Not sure if you have plans to add that in the future, but if not, I think it would be good to add it with this PR.

Good pointer. I've already started working on the next PR, so I'll attach this "addition" on that new PR.

RivasCVA commented:
> We need to validate too if the initial value is valid (and if the `validation` prop is set).

BTW, you actually meant "if the initial value is **invalid**," right?

jandrade commented:
> > We need to validate too if the initial value is valid (and if the `validation` prop is set).
> 
> BTW, you actually meant "if the initial value is **invalid**," right? So that there is an error by default at the start.

that's right :)

RivasCVA commented:
The parent pull-request ([#1126]](https://github.com/Khan/wonder-blocks/pull/1126)) has been merged into `master`, but this branch (`wb-1052`) now has conflicts with the new base branch. These conflicts must be resolved before checks can complete on this pull-request.
